app/lib/logic/provider_signup.rb

```diff
@@ -10,90 +10,6 @@ def signup_provider_possible?
       end
-
-      def signup_provider(application_plan = nil, options = {})
-        ensure_master
-
-        # using the buyers (instead of providers) and users (instead of admins) as mass assignment with scope throws errors
-        # either way, we're manually setting those fields
-        provider = self.buyers.build_with_fields
-        provider.provider = true
-        provider.validate_fields!
-        provider.sample_data = true
-        provider.signup_mode! # don't create service or app; validate subdomain instead of domain
-        provider.email_all_users = true
-
-        user = provider.users.build_with_fields
-        user.account = provider # this is being assigned manually as it's not set otherwise, investigate
-        user.role = :admin
-        user.username = "admin"
-        user.validate_fields!
-
-        impersonation_admin = ::Signup::ImpersonationAdminBuilder.build(account: provider)
-
-        yield(provider, user) if block_given?
-
-        service = services.default or raise "Missing default service of master"
-        application_plan ||= service.application_plans.default or raise "Missing Application plan of master"
-        account_plan = account_plans.default or raise "Missing default Account plan of master"
-        service_plan = service.service_plans.default or raise "Missing default Service plan of master"
-
-        return if [provider, user, impersonation_admin].any?(&:invalid?)
-
-        case provider.classify
-        when 'Internal'.freeze
-          provider.settings.assign_attributes(monthly_billing_enabled: false,
-                                              monthly_charging_enabled: false)
-        end
-
-        #  TODO: concurrency issue (already had it on remonster -
-        #  provider or user can become invalid during the process
-        #
-        self.class.transaction do
-          provider.save!
-
-          # trigger like this does not work in MySQL...
-          provider.update_attribute :tenant_id, provider.id
-
-          # ...since the other one wasn't in place on insertion
-          user.update_attribute :tenant_id, provider.id
-          impersonation_admin.update_attribute :tenant_id, provider.id
-          provider.fields_definitions.update_all "tenant_id = #{provider.id}"
-
-          account_plan.create_contract_with! provider
-          service_plan.create_contract_with! provider
-          application_plan.create_contract_with! provider
-
-          ServiceCreator.new(service: provider.services.build(name: 'API')).call(private_endpoint: BackendApi.default_api_backend)
-        end
-
-        if ThreeScale.config.onpremises
-          provider.force_upgrade_to_provider_plan!(application_plan)
-        else
-          ::Signup::ProviderAccountManager.set_provider_constraints(provider, application_plan)
-        end
-
-        SignupWorker.enqueue(provider)
-
-        third_party_notifications!(provider, options) unless options[:skip_third_party_notifications]
```

Comment:
I guess this condition was here only because of the tests, right?
How are the "third party notifications" (user signup tracking) handled now while using `Signup::ProviderAccountManager`?

Reply:
> I guess this condition was here only because of the tests, right?

Right :smile: 

> How are the "third party notifications" (user signup tracking) handled now while using Signup::ProviderAccountManager?

It's a mess :scream: Sometimes it is inside the controller, like in SignupsController#create, and sometimes it is inside `/lib/tracking_notifications.rb`. Before this PR it was also here in ProviderSignup.
And refactoring this is not trivial at all. The params sometimes are different. And even the user initialized in `/lib/tracking_notifications.rb` is the `current_user` because there it mostly tracks the login. Besides, there are some methods related to analytics tracking that I think that we do not use anymore but I didn't touch just in case.

I've just made another commit because you are right that we were introducing a bug by one of the 2 controllers that I've touched but I've only added it in the controller for simplicity and I think that we should create a Jira for refactoring all this analytics tracking mess. Maybe for example in a background job called from the batch of SignupWorker.

Let's talk about this tomorrow :+1: